src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -12268,5 +12332,123 @@ public static void Main()
+        D d = (int x = num + 3) => x;
+        Console.WriteLine(d());
+    }
+}
+""";
+            CompileAndVerify(source, expectedOutput: "7");
+        }
+
+        [Fact]
+        public void ImplicitLambdaDefaultParameter_NamedDelegateConversion()
+        {
+            var source = """
+using System;
+
+class Program
+{
+    delegate int D(int x = 3);
+    public static void Main()
+    {
+        D d = (x = 3) => x;
+        Console.WriteLine(d());
+    }
+}
+""";
+            CreateCompilation(source).VerifyDiagnostics(
+                    // (8,16): error CS9065: Implicitly typed lambda parameter 'x' cannot have a default value.
+                    //         D d = (x = 3) => x;
+                    Diagnostic(ErrorCode.ERR_ImplicitlyTypedDefaultParameter, "x").WithArguments("x").WithLocation(8, 16));
+        }
+
+        [Fact]
+        public void LambdaDefaultDiscardParameter_DelegateConversion_OptionalRequiredMismatch()
+        {
+            var source = """
+using System;
+
+class Program
+{
+    delegate int D(int x, int y);
+    public static void Main()
+    {
+        D d = (int _, int _ = 3) => 10;
+        Console.WriteLine(d(4));
+    }
+}
+""";
+            CreateCompilation(source).VerifyDiagnostics(
+                    // (8,27): error CS9067: Parameter 2 has default value '3' in lambda and '<missing>' in the target delegate type.
+                    //         D d = (int _, int _ = 3) => 10;
+                    Diagnostic(ErrorCode.ERR_OptionalParamValueMismatch, "_").WithArguments("2", "3", "<missing>").WithLocation(8, 27),
+                    // (9,27): error CS7036: There is no argument given that corresponds to the required formal parameter 'y' of 'Program.D'
+                    //         Console.WriteLine(d(4));
+                    Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "d").WithArguments("y", "Program.D").WithLocation(9, 27));
+        }
+
+        [Fact]
+        public void LambdaDefaultDisardParameter_DelegateConversion_DefaultValueMismatch()
+        {
+            var source = """
+using System;
+
+class Program
+{
+    delegate int D(int x, int y = 7);
+    public static void Main()
+    {
+        D d = (int _, int _ = 3) => 10;
+        Console.WriteLine(d(4));
+    }
+}
+""";
+            CreateCompilation(source).VerifyDiagnostics(
+                    // (8,27): error CS9067: Parameter 2 has default value '3' in lambda and '7' in the target delegate type.
+                    //         D d = (int _, int _ = 3) => 10;
+                    Diagnostic(ErrorCode.ERR_OptionalParamValueMismatch, "_").WithArguments("2", "3", "7").WithLocation(8, 27));
+        }
+
+        [Fact]
+        public void LambdaDefaultParameter_TargetTypeConversionWarning_ErrorInLambdaBody()
+        {
+            var source = """
+using System;
+
+class Program
+{
+    delegate void D(int x, int y);
+    public static void Main()
+    {
+        D d = (int x, int y = 4) => {
+            string s = 5;
```

Comment:
>string s = 5; [](http://example.com/codeflow?start=12&length=13)

Should we report an error for this assignment?

Reply:
Actually, it looks like we don't bind the lambda body if there is no conversion to the target delegate type.

Reply:
The current behavior for other target-type conversion errors is to avoid binding the body, so it seemed best to be consistent with what they do: see https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzguATEBqAPgAQCYBGAWACh8BmAAiJoGEaBvCm9mjAUwBsuBzCAi41YCGgBEAFGJoAPAJQBuNh1XsihFuo7sJnGgF4aU/MQAM8hUYB828rsccA9M5oIAngAcRAYwD2MABuXABOKFCBNGGh/qE6ThoWNChGNAAsKg6JugmJrjQAZv7+nP5cKDAA5OJcclBoeU4AYiVFbcYwXADuNK3+UspNNAC+WbojFCNAA